tests/ops/qutrit/test_qutrit_matrix_ops.py

```diff
@@ -283,9 +283,363 @@ def test_matrix_representation(self):
+        [
+            ([0, 1], 2, "ab", "String of control values can contain only '0' or '1' or '2'."),
+            ([0, 1], 2, "012", "Length of control trit string must equal number of control wires."),
+            ([0, 1], 2, [0, 1], "Alternative control values must be passed as a ternary string."),
+        ],
+    )
+    def test_invalid_mixed_polarity_controls(
+        self, control_wires, wires, control_values, expected_error_message
+    ):
+        """Test if ControlledQutritUnitary properly handles invalid mixed-polarity
+        control values."""
+        target_wires = Wires(wires)
+
+        with pytest.raises(ValueError, match=expected_error_message):
+            qml.ControlledQutritUnitary(
+                U_thadamard_01,
+                control_wires=control_wires,
+                wires=target_wires,
+                control_values=control_values,
+            ).matrix()
+
+    mixed_polarity_data = [
+        ([0], 1, "0"),
+        ([0], 1, "1"),
+        ([0], 1, "2"),
+        ([0, 1], 2, "00"),
+        ([0, 1], 2, "01"),
+        ([0, 1], 2, "02"),
+        ([1, 0], 2, "10"),
+        ([0, 1], 2, "11"),
+        ([0, 1], 2, "12"),
+        ([0, 1], 2, "20"),
+        ([1, 0], 2, "21"),
+        ([1, 0], 2, "22"),
+        ([0, 1], [2, 3], "01"),
+        ([0, 2], [3, 1], "12"),
+        ([1, 2, 0], [3, 4], "012"),
+        ([1, 0, 2], [4, 3], "210"),
+    ]
+
+    @pytest.mark.parametrize("control_wires,wires,control_values", mixed_polarity_data)
+    def test_mixed_polarity_controls(self, control_wires, wires, control_values, tol):
+        """Test if ControlledQutritUnitary properly applies mixed-polarity
+        control values."""
+        target_wires = Wires(wires)
+
+        dev = qml.device("default.qutrit", wires=len(control_wires + target_wires))
+
+        # Pick a random unitary
+        U = unitary_group.rvs(3 ** len(target_wires), random_state=1967)
+
+        # Pick random starting state for the control and target qutrits
+        starting_state = np.random.normal(size=3 ** (len(control_wires + target_wires)))
+        starting_state = starting_state / np.linalg.norm(starting_state)
+        print(np.linalg.norm(starting_state))
+        dev._state = starting_state.reshape([3] * len(control_wires + target_wires))
+        dev.apply(
+            [
+                qml.ControlledQutritUnitary(
+                    U,
+                    control_wires=control_wires,
+                    wires=target_wires,
+                    control_values=control_values,
+                ),
+            ]
+        )
+        final_state = dev._state.flatten()
+
+        dev.reset()
+        dev._state = starting_state.reshape([3] * len(control_wires + target_wires))
+        one_locations = [x for x in range(len(control_values)) if control_values[x] == "1"]
+        zero_locations = [x for x in range(len(control_values)) if control_values[x] == "0"]
+
+        ops = []
```

Comment:
I think I followed what's happening here (you are manually applying shifts to set control values, apply the operation, and then undoing them)? It would be helpful to add some developer-facing comments to explain the process a little more.